src/Workspaces/Core/Portable/Log/FileLogger.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Internal.Log
+{
+    /// <summary>
+    /// A logger that publishes events to a log file.
+    /// </summary>
+    internal sealed class FileLogger : ILogger
+    {
+        private readonly object _gate;
+        private readonly string _logFilePath;
+        private readonly StringBuilder _buffer;
+        private bool _enabled;
+
+        public FileLogger(IGlobalOptionService optionService, string logFilePath)
+        {
+            _logFilePath = logFilePath;
+            _gate = new();
+            _buffer = new();
+            _enabled = optionService.GetOption(InternalDiagnosticsOptions.EnableFileLoggingForDiagnostics);
+            optionService.OptionChanged += OptionService_OptionChanged;
+        }
+
+        public FileLogger(IGlobalOptionService optionService)
+            : this(optionService, Path.Combine(Path.GetTempPath(), "Roslyn", "Telemetry", GetLogFileName()))
+        {
+        }
+
+        private static string GetLogFileName()
+            => DateTime.Now.ToString().Replace(' ', '_').Replace('/', '_').Replace(':', '_') + ".log";
+
+        private void OptionService_OptionChanged(object? sender, OptionChangedEventArgs e)
+        {
+            if (e.Option == InternalDiagnosticsOptions.EnableFileLoggingForDiagnostics)
+            {
+                Contract.ThrowIfNull(e.Value);
+
+                _enabled = (bool)e.Value;
+            }
+        }
+
+        public bool IsEnabled(FunctionId functionId)
+        {
+            if (!_enabled)
+            {
+                return false;
+            }
+
+            // Limit logged function IDs to keep a reasonable log file size.
+            var str = functionId.ToString();
+            return str.StartsWith("Diagnostic") ||
+                str.StartsWith("CodeAnalysisService") ||
+                str.StartsWith("Workspace") ||
+                str.StartsWith("WorkCoordinator") ||
+                str.StartsWith("IncrementalAnalyzerProcessor") ||
+                str.StartsWith("ExternalErrorDiagnosticUpdateSource");
+        }
+
+        private void Log(FunctionId functionId, string message)
```

Comment:
>Log [](http://example.com/codeflow?start=21&length=3)

Is this supposed to perform blocking IO?

Reply:
This will only be enabled when the user has turned on the option to collect logs, so I thought it’s fine. I can change this logic if you strongly feel otherwise.

Reply:
I moved the logging/IO to a separate task queue so the Log function is no longer blocking.